Admin/Test-AMSI.ps1

```diff
@@ -1,144 +1,782 @@
+        }
     }
-    if ($CheckAMSIProviders) {
-        $AMSI = Get-ChildItem 'HKLM:\SOFTWARE\Microsoft\AMSI\Providers' -Recurse
-        $AMSI -match '[0-9A-Fa-f\-]{36}' | Out-Null
-        $Matches.Values | ForEach-Object { Get-ChildItem "HKLM:\SOFTWARE\Classes\ClSid\{$_}" | Format-Table -AutoSize }
+
+    function GetWindowsMayorVersion {
+        param(
+            [Parameter(Mandatory = $true)]
+            [string]$ExchangeServer
+        )
+
+        if ( $ExchangeServer.ToLower() -eq $env:COMPUTERNAME.ToLower() ) {
+            [System.Environment]::OSVersion.Version.Major
+        } else {
+            $temp = $null
+            $temp = Get-CimInstance -ComputerName $ExchangeServer -ClassName Win32_OperatingSystem -ErrorAction SilentlyContinue
+            if ( $null -eq $temp ) {
+                $temp = Get-WmiObject -ComputerName $ExchangeServer -Query 'Select Version from Win32_OperatingSystem' -ErrorAction SilentlyContinue
+            }
+            if ( $temp ) {
+                $temp.Version.Split('.')[0]
+            } else {
+                0
+            }
+        }
     }
-    if ($EnableAMSI) {
-        Remove-SettingOverride -Identity DisablingAMSIScan -Confirm:$false
-        Get-ExchangeDiagnosticInfo -Process Microsoft.Exchange.Directory.TopologyService -Component VariantConfiguration -Argument Refresh
-        Write-Warning "Remember to restart IIS for this to take affect. You can accomplish this by running .\Test-AMSI.ps1 -RestartIIS"
-        return
+
+    function CheckAMSIProviders {
+        param(
+            [Parameter(Mandatory = $true)]
+            [string]$ExchangeServer
+        )
+
+        $AMSIProviders = $null
+        Write-Output $msgNewLine
+        if ( $ExchangeServer.ToLower() -eq $env:COMPUTERNAME.ToLower() ) {
+            Write-Host "Checking local AMSI Provider on $ExchangeServer"
+            $AMSIProviders = Get-ChildItem 'HKLM:\SOFTWARE\Microsoft\AMSI\Providers' -Recurse -ErrorAction SilentlyContinue
+        } else {
+            Write-Host "Checking remote AMSI Provider on $ExchangeServer"
+            $AMSIProviders = Invoke-Command -ComputerName $ExchangeServer -ScriptBlock { Get-ChildItem 'HKLM:\SOFTWARE\Microsoft\AMSI\Providers' -Recurse -ErrorAction SilentlyContinue } -ErrorAction SilentlyContinue -ErrorVariable InvokeError
+            if ( $InvokeError ) {
+                Write-Warning "We could not connect with server $ExchangeServer"
+                return
+            }
+        }
+        if ( $null -eq $AMSIProviders ) {
+            Write-Warning "We did not find any AMSI Provider"
+        } else {
+            foreach ( $provider in $AMSIProviders) {
+                $provider -match '[0-9A-Fa-f\-]{36}' | Out-Null
+                foreach ($m in $Matches.Values ) {
+                    $key = "HKLM:\SOFTWARE\Classes\ClSid\{$m}"
+                    Write-Output $msgNewLine
+                    if ( $ExchangeServer.ToLower() -eq $env:COMPUTERNAME.ToLower() ) {
```

Comment:
Utilize the shared function of `Invoke-ScriptBlockHandler` in `Shared` directory.